sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala

```diff
@@ -217,16 +234,17 @@ class UnivocityParser(
           timestampFormatter.parse(datum)
         } catch {
           case NonFatal(e) =>
-            // If fails to parse, then tries the way used in 2.0 and 1.x for backwards
-            // compatibility.
-            val str = DateTimeUtils.cleanLegacyTimestampStr(UTF8String.fromString(datum))
-            DateTimeUtils.stringToTimestamp(str, options.zoneId).getOrElse {
-              // There may be date type entries in timestamp column due to schema inference
-              if (options.inferDate) {
-                daysToMicros(dateFormatter.parse(datum), options.zoneId)
-              } else {
-                throw(e)
+            // There may be date type entries in timestamp column due to schema inference
+            if (options.inferDate) {
```

Comment:
This reorganisation is required to define priority of parsing timestamps as dates when `inferDate = true`. Since this is optional, we try this first and then fall back to the legacy behaviour if it is enabled.

Reply:
cc @cloud-fan @Jonathancui123

Reply:
Thanks for the fix! It also makes sense that you have reverted "fix tests for SPARK-39469". 

This implements @cloud-fan's suggestion and the "Option A" that I described [in my comment](https://github.com/apache/spark/pull/36871#issuecomment-1192923903) which considers type coercion between Dates and Timestamps. I'm okay with this - as a user, it makes sense that when I can turn on `inferDate`, a date entry is detected in a timestamp column and then cast to a timestamp. 

However, I think @HyukjinKwon should take a look. He [said that it would be more future-proof](https://github.com/apache/spark/pull/36871#issuecomment-1193016381) to not provide this functionality.

Reply:
Thanks 👍. Yeah, it would be good if Hyukjin could take another look.

Reply:
It's a small issue within the fallback branch and is probably not an issue without this PR that tries to fix the fallback branch. I'm OK to not open a new PR for it.